File: app/src/main/java/com/slidely/photoframe/repository/DefaultPreferenceRepository.java
package com.slidely.photoframe.repository;

import android.app.Activity;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.slidely.photoframe.R;
import com.slidely.photoframe.features.preferences.PreferencesFragment;

import java.util.concurrent.TimeUnit;

public class DefaultPreferenceRepository implements PreferenceRepository {
    private final Activity activity;

    DefaultPreferenceRepository(Activity activity) {
        this.activity = activity;
    }

    public int get() {
        SharedPreferences speedPreference = PreferenceManager.getDefaultSharedPreferences(activity);
        int defSpeedValue = activity.getResources().getInteger(R.integer.speed_default);
        return (int) (Integer.parseInt(speedPreference.getString(PreferencesFragment.SPEED_KEY,
                "" + defSpeedValue)) * TimeUnit.SECONDS.toMillis(1));
    }
}

File: app/src/main/java/com/k4dima/photoframe/features/main/data/DefaultPreferenceRepository.kt
package com.k4dima.photoframe.features.main.data

import android.content.Context
import android.preference.PreferenceManager
import com.f2prateek.rx.preferences2.RxSharedPreferences
import com.k4dima.photoframe.R
import com.k4dima.photoframe.features.main.ui.di.MainScope
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@MainScope
class DefaultPreferenceRepository
@Inject
internal constructor(private val context: Context) : PreferenceRepository {
    companion object {
        internal fun speedKey(context: Context) = context.getString(R.string.speed_key)!!
    }

    private val preferencesSingle = Single.fromCallable { PreferenceManager.getDefaultSharedPreferences(context) }
            .subscribeOn(Schedulers.io())
    /** Keep a strong reference on your RxSharedPreferences instance for as long as you want to observe them
     * to prevent listeners from being GCed.*/
    private lateinit var rxPreferences: RxSharedPreferences
    private var rxPreferencesSingle = preferencesSingle.map {
        RxSharedPreferences.create(it).apply { rxPreferences = this }
    }!!
    override val fullscreen = preferencesSingle
            .map { it.getBoolean(context.getString(R.string.fullscreen_key), true) }!!
    override val clock = rxPreferencesSingle.map { it.getBoolean(context.getString(R.string.clock_key), false) }
            .toObservable()
            .flatMap { it.asObservable() }!!
    override val sound = preferencesSingle.map { it.getBoolean(context.getString(R.string.sound_key), false) }!!
    /**In milliseconds*/
    override val speed: Single<Int> = preferencesSingle.map {
        val defaultSpeedValue = "" + context.resources.getInteger(R.integer.speed_default)
        (Integer.parseInt(it.getString(speedKey(context),
                defaultSpeedValue)) * TimeUnit.SECONDS.toMillis(1)).toInt()
    }
}